include/modem/modem_slm.h

```diff
@@ -0,0 +1,203 @@
+
+#ifndef MODEM_SLM_H_
+#define MODEM_SLM_H_
+
+/**
+ * @file modem_slm.h
+ *
+ * @defgroup modem_slm Modem SLM library
+ *
+ * @{
+ *
+ * @brief Public APIs for the Modem SLM library.
+ */
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <stddef.h>
+#include <stdbool.h>
+#include <zephyr/types.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/util_macro.h>
+#include <zephyr/toolchain/common.h>
+
+/** Max size of AT command response is 2100 bytes. */
+#define SLM_AT_CMD_RESPONSE_MAX_LEN 2100
+
+/**
+ * @brief AT command result codes
+ */
+enum at_cmd_state {
+	AT_CMD_OK,
+	AT_CMD_ERROR,
+	AT_CMD_ERROR_CMS,
+	AT_CMD_ERROR_CME,
+	AT_CMD_PENDING
+};
+
+/**
+ * @typedef slm_data_handler_t
+ *
+ * Handler to handle data received from SLM, which could be AT response, AT notification
+ * or or simply raw data (e.g. DFU image).
+ *
+ * @param data    Data received from SLM.
+ * @param datalen Length of the data received.
+ *
+ * @note The handler runs from uart callback. It must not call @ref modem_slm_send_cmd. The data
+ * should be copied out by the application as soon as called.
+ */
+typedef void (*slm_data_handler_t)(const uint8_t *data, int datalen);
+
+/**
+ * @typedef slm_ind_handler_t
+ *
+ * Handler to handle MODEM_SLM_INDICATE_PIN signal from SLM.
+ */
+typedef void (*slm_ind_handler_t)(void);
+
+/**@brief Initialize Modem SLM library.
+ *
+ * @param handler Pointer to a handler function of type @ref slm_data_handler_t.
+ *
+ * @return Zero on success, non-zero otherwise.
+ */
+int modem_slm_init(slm_data_handler_t handler);
+
+/**@brief Un-initialize Modem SLM library.
+ *
+ */
+int modem_slm_uninit(void);
+
+/**
+ * @brief Register callback for MODEM_SLM_INDICATE_PIN indication
+ *
+ * @param handler Pointer to a handler function of type @ref slm_ind_handler_t.
+ * @param wakeup  Enable/disable System Off wakeup by GPIO Sense.
+ *
+ * @retval Zero    Success.
+ * @retval -EFAULT if MODEM_SLM_INDICATE_PIN is not defined.
+ */
+int modem_slm_register_ind(slm_ind_handler_t handler, bool wakeup);
+
+/**
+ * @brief Wakeup nRF9160 SiP via MODEM_SLM_WAKEUP_PIN
+ *
+ * @return Zero on success, non-zero otherwise.
+ */
+int modem_slm_wake_up(void);
+
+/**
+ * @brief Reset the RX function of the serial interface
+ *
+ */
+void modem_slm_reset_uart(void);
+
+/**
+ * @brief Function to send an AT command in SLM command mode
+ *
+ * This function wait until command result is received. The response of the AT command is received
+ * via the @ref slm_ind_handler_t registered in @ref modem_slm_init.
+ *
+ * @param command Pointer to null terminated AT command string without command terminator
+ * @param time_ms Timeout for the command in milliseconds, Zero means infinite wait
+ *
+ * @retval success  Zero or positive @ref at_cmd_state if command execution succeeds.
+ * @retval -EAGAIN if command execution times out.
+ * @retval other   negative value if command execution fails.
+ */
+int modem_slm_send_cmd(const char *const command, int time_ms);
```

Comment:
I think `timeout_ms` would be more descriptive for the timeout parameter.

Reply:
yes, and I realize that we should use seconds instead of milliseconds for timeout